### PR TITLE
curvefs/metasever: correct misunderstood of index in s3objname

### PR DESCRIPTION
**Pull request by h0hmj**
current mapping from (off, len) to objname is wrong, index in s3objname is wrong.

wrong understanding: a s3chunk(chunkid) is split in to several fixed sized s3objs, and index them from 0
correct understanding: index = offset_in_a_chunkindex/s3_object_max_size

### CODE REVIEW

File: curvefs/conf/metaserver.conf [hunk 1]
```diff
@@ -35,7 +35,7 @@ s3compactwq.queue_size=5
 s3compactwq.fragment_threshold=20
 # max chunks to process per compact task
 s3compactwq.max_chunks_per_compact=10
-s3compactwq.enqueue_sleep_ms=100
+s3compactwq.enqueue_sleep_ms=1000
```
Comment (wu-hanqing) [inline curvefs/conf/metaserver.conf:38]: add a description about this item

File: curvefs/src/metaserver/s3compact_wq_impl.cpp [hunk 2]
```diff
@@ -327,8 +378,21 @@ void S3CompactWorkQueueImpl::CompactChunks(
     PartitionInfo pinfo, std::shared_ptr<CopysetNodeWrapper> copysetNodeWrapper,
     std::shared_ptr<S3AdapterManager> s3adapterManager,
     std::shared_ptr<S3InfoCache> s3infoCache) {
+    auto cleanup = absl::MakeCleanup([this, inodeKey]() {
+        std::unique_lock<std::mutex> guard(mutex_);
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/s3compact_wq_impl.cpp:382]: std::lock_guard is enough

File: curvefs/src/metaserver/s3compact_wq_impl.cpp [hunk 3]
```diff
@@ -120,14 +124,15 @@ S3CompactWorkQueueImpl::BuildValidList(const S3ChunkInfoList& s3chunkinfolist,
                                        uint64_t inodeLen) {
     std::list<struct S3CompactWorkQueueImpl::Node> validList;
     std::function<void(uint64_t, uint64_t, uint64_t, uint64_t, uint64_t,
-                       uint64_t)>
+                       uint64_t, uint64_t)>
```
Comment (wu-hanqing) [inline curvefs/src/metaserver/s3compact_wq_impl.cpp:127]: does these seven parameters are corresponding to fields in `S3CompactWorkQueueImpl::Node`, if so, why not use Node directly?

File: curvefs/conf/metaserver.conf (hunk 1 above)
Comment (h0hmj) [reply curvefs/conf/metaserver.conf:39]: fix

File: curvefs/src/metaserver/s3compact_wq_impl.cpp (hunk 2 above)
Comment (h0hmj) [reply curvefs/src/metaserver/s3compact_wq_impl.cpp]: fix

File: curvefs/src/metaserver/s3compact_wq_impl.cpp (hunk 3 above)
Comment (h0hmj) [reply curvefs/src/metaserver/s3compact_wq_impl.cpp]: good idea. fixed.

**Review by wu-hanqing — APPROVED** [review]

**Comment by h0hmj** [comment]
recheck

**Comment by h0hmj** [comment]
recheck

File: curvefs/src/metaserver/s3compact_wq_impl.cpp [hunk 4]
```diff
@@ -216,55 +225,72 @@ S3CompactWorkQueueImpl::BuildValidList(const S3ChunkInfoList& s3chunkinfolist,
...
             curr = next;
             continue;
         }
-        // block index is relatively in a s3chunkinfo
-        for (uint64_t index = (curr->begin - curr->chunkoff) / blockSize;
-             index * blockSize + curr->chunkoff < curr->end; index++) {
+
+        VLOG(9) << "s3compact: read [" << curr->begin << "-" << curr->end
+                << "]";
+        uint64_t startIndex =
```
Comment (baijiaruo) [inline curvefs/src/metaserver/s3compact_wq_impl.cpp:241]: curr->begin - curr->begin / chunkSize * chunkSize is equivalent to curr->begin - curr->begin / chunkSize * chunkSize
startIndex = begin % chunksize / blocksize?

File: curvefs/src/metaserver/s3compact_wq_impl.cpp [hunk 5]
```diff
@@ -216,55 +225,72 @@ S3CompactWorkQueueImpl::BuildValidList(const S3ChunkInfoList& s3chunkinfolist,
...
+            (curr->begin - curr->begin / chunkSize * chunkSize) / blockSize;
+        for (uint64_t index = startIndex;
+             curr->begin / chunkSize * chunkSize + index * blockSize <=
```
Comment (baijiaruo) [inline curvefs/src/metaserver/s3compact_wq_impl.cpp:244]: The comparison of for loops is as concise as possible

File: curvefs/src/metaserver/s3compact_wq_impl.cpp [hunk 6]
```diff
@@ -216,55 +225,72 @@ S3CompactWorkQueueImpl::BuildValidList(const S3ChunkInfoList& s3chunkinfolist,
 
 int S3CompactWorkQueueImpl::ReadFullChunk(
     const std::list<struct S3CompactWorkQueueImpl::Node>& validList,
-    uint64_t fsId, uint64_t inodeId, uint64_t blockSize, std::string* fullChunk,
-    uint64_t* newChunkid, uint64_t* newCompaction, S3Adapter* s3adapter) {
+    uint64_t fsId, uint64_t inodeId, uint64_t blockSize, uint64_t chunkSize,
+    std::string* fullChunk, uint64_t* newChunkid, uint64_t* newCompaction,
+    S3Adapter* s3adapter) {
```
[the remark this answers is not in the document]
Comment (h0hmj) [reply curvefs/src/metaserver/s3compact_wq_impl.cpp:230]: todo, maybe next time. struct a ctx for task

File: curvefs/src/metaserver/s3compact_wq_impl.cpp (hunk 4 above)
Comment (h0hmj) [reply curvefs/src/metaserver/s3compact_wq_impl.cpp]: add posRoundDown = pos/chunkSize*chunkSize to simplify code. it is reused many times.
%  = pos - posRoundDown

File: curvefs/src/metaserver/s3compact_wq_impl.cpp (hunk 5 above)
Comment (h0hmj) [reply curvefs/src/metaserver/s3compact_wq_impl.cpp]: same as above

**Comment by h0hmj** [comment]
recheck

**Review by baijiaruo — APPROVED** [review]

**Comment by h0hmj** [comment]
recheck

**Pull request merged**